core/src/main/java/com/linecorp/armeria/client/cookie/DefaultCookieJar.java

```diff
@@ -90,16 +90,31 @@ public void set(URI uri, Iterable<? extends Cookie> cookies, long createdTimeMil
         lock.lock();
         try {
             for (Cookie cookie : cookies) {
-                cookie = ensureDomainAndPath(cookie, uri);
+                final Cookie ensuredCookie = ensureDomainAndPath(cookie, uri);
                 // remove similar cookie if present
-                store.removeLong(cookie);
-                if ((cookie.maxAge() == Cookie.UNDEFINED_MAX_AGE || cookie.maxAge() > 0) &&
-                    cookiePolicy.accept(uri, cookie)) {
-                    store.put(cookie, createdTimeMillis);
-                    final Set<Cookie> cookieSet = filter.computeIfAbsent(cookie.domain(), s -> new HashSet<>());
-                    // remove similar cookie if present
-                    cookieSet.remove(cookie);
-                    cookieSet.add(cookie);
+                store.removeLong(ensuredCookie);
+                if ((ensuredCookie.maxAge() == Cookie.UNDEFINED_MAX_AGE || ensuredCookie.maxAge() > 0) &&
+                    cookiePolicy.accept(uri, ensuredCookie)) {
+                    store.put(ensuredCookie, createdTimeMillis);
+                    final Set<Cookie> cookieSet = filter.computeIfAbsent(ensuredCookie.domain(),
+                                                                         s -> new HashSet<>());
+
+                    // remove the cookie with the same name, domain, and path if present
+                    // https://datatracker.ietf.org/doc/html/rfc6265#page-24
+                    final Cookie oldCookie = cookieSet.stream()
+                                                      .filter(c -> isSameNameAndPath(c, ensuredCookie))
+                                                      .findFirst()
+                                                      .orElse(null);
+                    if (oldCookie == null) {
+                        cookieSet.add(ensuredCookie);
+                        continue;
+                    }
+                    if (!uri.getScheme().startsWith("http") && oldCookie.isHttpOnly()) {
```

Comment:
Should we compare the entire value instead of using `startWith()` 
```suggestion
                    if (!(uri.getScheme().equals("http") ||
                          uri.getScheme().equals("https")) && 
                         oldCookie.isHttpOnly()) {
```